include/evmc/evmc.h

```diff
@@ -141,6 +141,18 @@ struct evmc_message
      * Ignored in evmc_execute_fn().
      */
     evmc_bytes32 create2_salt;
+
+    /**
+     * The address of the code to be executed.
+     *
+     * May be different from the evmc_message::destination (recipient) in case of
+     * ::EVMC_CALLCODE or ::EVMC_DELEGATECALL.
```

Comment:
@yperbasis, When working on EVMC compatibility in [Nimbus-eth1](https://github.com/status-im/nimbus-eth1), I was surprised to find that correct handling of `msg.destination` received via `evmc_call_fn` required checking whether `msg.destination` corresponds to a valid precompile in the current block's fork.  This is because `DELEGATECALL` and `CALLCODE` lose their special behaviour when calling a precompile.

Perhaps the EVMC API comment should mention that it depends on whether the destination is a precompile?

Reply:
I don't believe this is the case. What is so special about precompiles in this context?

Reply:
Precompiles are handled on the Host side, not VM side (in EVMC terms). After this change Hosts should use `msg.code_address` rather than `msg.destination` to determine which smart contract or precompile to execute. On the VM side DELEGATECALL and CALLCODE shouldn't have any special behaviour depending on whether the code address is a precompile or not. At least in EVMC terms, where VM is not responsible for things like maintaining EIP-2929 `accessed_addresses`.

For example, that's how the Host side looks in Silkworm after this change: https://github.com/torquem-ch/silkworm/blob/master/core/silkworm/execution/evm.cpp#L156

Reply:
@chfast, before this change,`msg.destination` given to `evmc_call_fn` always contained the address for looking up the contract bytecode for all types of call (obviously not for precompiles), and yes of course that's a host side operation.

Then for `DELEGATECALL` and `CALLCODE`, `msg.destination` had to be replaced with the recipient address before calling `evmc_execute_fn`.

Silkworm calculated the recipient [here](https://github.com/torquem-ch/silkworm/blob/27a26b062652af38cd35c77ba3c86ed8077d01e4/core/silkworm/execution/evm.cpp#L227), and then overwrotes `msg.destination` [here](https://github.com/torquem-ch/silkworm/blob/27a26b062652af38cd35c77ba3c86ed8077d01e4/core/silkworm/execution/evm.cpp#L212) *conditional on the `evmc_execute_fn` not being a precompile call*.

In Nimbus I simplified that to just overwrite `msg.destination` with the equivalent of `address_stack_.top()` if doing `DELEGATECALL` or `CALLCODE` (to a non-precompile) before calling `evmc_execute_fn`, which works out the same as Silkworm (because it is equal to `msg.sender` in the `CALLCODE` case).

Now, after this change, `msg.destination` no longer needs overwriting by the host, so the meaning of `msg.destination` is *no longer* dependent on whether it's a fork-specific precompile address.

So I retract my suggestion because I'd mistakenly thought the only change was adding "missing but calculatable" field `msg.code_address`.  I hadn't realised the meaning of `msg.destination` has also been changed.

ps. It is neat that `msg` can now be passed along unchanged by nested calls Until now I thought it had been an intentional design hack to overload `msg.destination` in this clever way to save a field. That proved quite handy to make me think harder about the information flows in Nimbus and how the EVM isn't allowed to know the code address (ironically, in practice it will now be passed the code address, though it's not supposed to use it).

Reply:
(I've edited my comment above; if you read the old version and ignored it, please read again, there remains an unresolved issue with regard to the new `msg.code_address` change and `EVMC_CAPABILITY_PRECOMPILES`.

Reply:
@yperbasis,
> Precompiles are handled on the Host side, not VM side (in EVMC terms).

Nimbus has called precompiles via EVMC since it first supported any kind of EVMC a couple of years ago, and the flag`EVMC_CAPABILITY_PRECOMPILES` was added since then, but we didn't bother using it until recently.

If precompiles are _required_ to be handled on the host side, that's surprising news.

Reply:
Yes, I can see the problem now.
I need to track Host implementations if they change `msg.destination` on the way. Even if they do, that's probably not recommended thing to do and EVMC precompiles implementation should just only consider `msg.code_address` for the precompile selection. Is this a problematic change?

What I plan to do as a follow up is to clarify the documentation and change EVMC examples. Might be also nice to change `destination` to `recipient`. How about we discuss the problem there?

Reply:
I don't mind where the discussion takes place (though I tend to miss some GitHub notications without a poke).

Documentation on the flow from EVM caller to nested EVM callee would be a nice improvement.  It might be enough just to expand the comments a bit.  There were quite a few things I had to figure out.  I don't need that documentation any more, but it would make EVMC clearer for the next perosn.

I think all host implementations must have changed `msg.destination` (before `msg.code_address` was added), for regular, non-precompiles, to get the correct answers from `DELEGATECALL` and `CALLCODE`.  And since `msg.code_address` they must stop changing `msg.destination` to get the correct answers. So perhaps there's no need to track the implementations. If they run consensus tests, they'll notice.

That leaves only precompiles. I don't know if any implementations call precompiles via EVMC, or if any loadable EVM provides precompiles (please let me know if you find one). Google returns no interesting matches for `EVMC_CAPABILITY_PRECOMPILES`. Nimbus is on track to use it, though.

This change isn't a problem. As EVMC precompile calls are Nimbus to Nimbus only as far as I know, no problem changing it to `msg.code_address`. It would be good to have a clear resolution on what the API-correct thing is if only for "someday" compatibility with others. If nothing is really correct, we can switch to a vendor extension if that helps. The comment at `EVMC_CAPABILITY_PRECOMPILES` about EIP-1352 is wrong anyway, as the host must use the fork-specific threshold due to some consensus tests using the low address range.

Reply:
> If they run consensus tests, they'll notice.

No, wait, who am I kidding. If both the host and EVM are doing the same wrong thing with `msg.code_address` added, they won't notice, and I'd guess cross-testing between different hosts and EVMs using the full consensus suite is not done much.

Reply:
> Nimbus has called precompiles via EVMC since it first supported any kind of EVMC a couple of years ago, and the flag`EVMC_CAPABILITY_PRECOMPILES` was added since then, but we didn't bother using it until recently.
> 
> If precompiles are _required_ to be handled on the host side, that's surprising news.

My bad, I forgot about `EVMC_CAPABILITY_PRECOMPILES`.

Reply:
See https://github.com/ethereum/evmc/pull/615.